feature: support customized sidebar item name from content

Jacco-V opened the pull request:
Add support for setting a customized sidebar item name from the documentation markdown content

See also issue: [#1821](/docsifyjs/docsify/issues/1821)

<!--
  PULL REQUEST TEMPLATE
  ---
  Please use English language
  Please don't delete this template
  ---
  Update "[ ]" to "[x]" to check a box in any list below.
  ---
  To avoid wasting your time, it's best to open a **feature request issue** first and wait for approval before working on it.
-->

## **Summary**

<!--
 THIS IS REQUIRED! Please describe what the change does and why it should be merged.
-->

This feature add support for specifying custom sidebar item names (especially for long titles) in the documentation markdown files, so that a the custom (typically shortened) item name is used in the sidebar:

```md
## How would I write an "hello, world" example? :sidebar="hello world?"
```

<!--
  If changing the UI in any way, please provide the a **before/after** screenshot:
-->

![image](https://user-images.githubusercontent.com/19333579/174432003-2e234fb2-3527-441c-87e0-e0244240be4a.png)

## **What kind of change does this PR introduce?**

<!--
  Copy/paste one of the following options:
-->

<!--
  Bugfix
  Feature
  Code style update
  Refactor
  Docs
  Build-related changes
  Repo settings
  Other
-->

Feature

<!--
  If you chose Other, please describe.
-->

## **For any code change,**

- [x] Related documentation has been updated if needed
- [x] Related tests have been updated or tests have been added

## **Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

## **Related issue, if any:**

<!-- Paste issue's link or number hashtag here. -->

## **Tested in the following browsers:**

- [x] Chrome
- [x] Firefox
- [x] Safari
- [x] Edge
- [ ] IE

julienw commented on src/core/render/compiler.js:
```diff
@@ -207,38 +207,36 @@ export class Compiler {
       _self.toc.push(nextToc);
 
-      return `<h${level} id="${slug}"><a href="${url}" data-id="${slug}" class="anchor"><span>${str}</span></a></h${level}>`;
+      return `<h${level} id="${slug}"><a href="${url}" data-id="${slug}" class="anchor"><span>${title}</span></a></h${level}>`;
```
I haven't tried the patch, but looking at this I would think that the sidebar text would also be used in the rendered page, but I don't think this is correct. We want the sidebar text in the sidebar but not in the page.

Have you tried this? :-)

Jacco-V replied:
No, I have not tried the changes, because I do not have a working frontend development environment. (I'm a backend dev giving this a try).

sy-records replied:
create a _sidebar.md https://docsify.js.org/#/more-pages?id=sidebar

Jacco-V commented:
I pushed some changes, but it does not work as expected :(
It seems that the `:sidebar="hello world?"` does not get parsed as `config`?

```md
## How would I write an "hello, world" example? :sidebar="hello world?"
```

Jacco-V commented:
I've added quoted string support to the [getAndRemoveConfig](https://github.com/Jacco-V/docsify/blob/feature-customizableSidebarItemName/src/core/render/utils.js#L21=L44) function and updated the tests.

Jacco-V commented:
Any chance anyone of the reviewers is going to review this?

Jacco-V commented:
**Code has been tested and is working as intended**, with the sole caveat that:

* we cannot support *escaped characters in quoted-string config-values*, because by the time the `str` is passed to the [`getAndRemoveConfig()`](https://github.com/docsifyjs/docsify/blob/da7937e340ec44b514fc397b3c1b1734eb938952/src/core/render/utils.js#L21=L44) method, it has been html-escaped, and backslashes have been stripped. I expect his will not be an issue for the large majority of use cases.

Koooooo-7 commented on src/core/render/utils.js:
```diff
@@ -23,16 +23,18 @@ export function getAndRemoveConfig(str = '') {
-        }
+      .replace(
+        /(?:^|\s):([\w-]+:?)=?([\w-%]+|&quot;((?!&quot;).)*&quot;|[“”][^“”]*[“”])?/g, // Note: because the provided `str` argument has been html-escaped, with backslashes stripped, we cannot support escaped characters in quoted strings :-(
+        function (m, key, value) {
```
FYI, there is a bit complicated regex including `negative lookahead`, we may need more tests here.

Jacco-V replied:
Yes, the regex became more complicated because the incomming `str` argument has been transformed from markdown to html. This is also the reason we cannot support escaped characters in the quoted-string config-values. It would have been a lot cleaner if the markdown to html transformation would have been applied after the config strings would have been read.

Jacco-V commented:
bump

Jacco-V commented:
Hi,
Is anyone going to review this PR? or is this functionality somehow not good enough?

Cheers,

Koooooo-7 reviewed (commented):
Hi @Jacco-V , Custom sidecar item name is a good purpose.
Personally, I m considering what is the good `mark` to define a config content , whether we should use the `"` here or not . therefore I may not approve this immediately.

For the instance, based on this PR , using the `mark` to include content means this `mark` can not be used. i.e.
If using `"` as  the mark ( `:sidebar="content"`), we can not use the `"` such as `:sidebar="the include "  content "` like you mentioned.
Instead, if we change the `mark` to `:sidebar={#"any content {in here}"}` like:
```js
    str = str
      .replace(
         /(?:^|\s):([\w-]+:?)=?([\w-%]+|{#.*})?/g,
        function (m, key, value) {
          if (key.indexOf(':') === -1) {
            config[key] = (value && value.replace(/{#|}$/g, '')) || true;
            return '';
          }

          return m;
        }
      )
      .replace(/^('|")|('|")$/g, '')
      .trim();
```
we can get this thing :
<img width="456" alt="image" src="https://user-images.githubusercontent.com/33706142/187407836-80aeb25f-23a0-47f8-a494-434f0be27314.png">

The point is what the `mark` should be the common rule/guidline for those exist/future config and it should not be very common use in user content.

Any ideas?

Jacco-V commented:
@Koooooo-7 Yes, I can see what you mean. Architecturally speaking, it would be best if the `str` argument would be passed to the `getAndRemoveConfig()` function _before_ the markdown has been transformed into html.

because, if this was the case, we could fully support quoted strings, including escape characters. This would mean you get the benefit of a using a syntax everybody knows, and knows what to expect from. 

```
:sidebar="In quoted strings we can use escaped \" characters"
```

However, I'm not sure it is possible to do this, without significant changes to the code base. Having said that, from an architectural point of view, it would still be the best option.

Another option would be to replace the regular-expression based 'parsing' with an actual parser, which would be able to handle escape charters even in the html formatted `str` argument. This option would get the quoted strings working, without addressing the architectural issue, but the end user won't see the difference.

The other option could be as you propose: go for a different set of 'start of string' and 'end of string' markers. This would sidestep the problem. The code-changes would be more simple, at the cost introducing a custom format instead of using quoted strings.

In the end, it is not up to me to choose which option is best suited for this project: I'm not one of the project members; I'm just a back-end guy who contributed a pull-request.

Koooooo-7 commented:
@Jacco-V  Thx for your input !
About change the `regex` to `parser` for configs is a good idea but we have not moved forward to it yet.
And exposing the `getAndRemoveConfig` as a hook is on the road. I suppose this hook could let user to custom config tpl more flexible, we could only provide a happy path one by default.

jhildenbiddle commented:
@Jacco-V @Koooooo-7 --

I haven't tested this PR, but perhaps we can work with the escaped character limitations using one of the following approaches:

1. Allow titles to be wrapped in single or double quotes. This allows for either quote type to be used within the string without needing to be escaped:
   ```
   :sidebar="Single 'quotes' here"
   ```
   ```
   :sidebar='Double "quotes" here'
   ```
2. Use HTML entities
   ```
   :sidebar='Double &quot;quotes&quot; here'
   ```


FWIW, I would also change the "Customize..." sidebar links to the following:

- Helpers > Heading IDs
- Helpers > Sidebar Titles

Koooooo-7 commented:
@jhildenbiddle How you doing ~ 

About the `Customize ` changes, I think it is a nice change to make the label more plain and understandable.

The `"/'` approach it sounds both fine. The more importance thing to me is the now and further enrichment configs, I think we should refactor `getAndRemoveConfig` to be  more flexible parser(s) for configs, otherwise , the regex would be more and more complex and messy.

Jacco-V commented:
@jhildenbiddle This is a possibility, and it could be an intermediate solution, but it would also be another incarnation of 'sidestepping the problem', although with a better solution than introducing custom `start of string` and `end of string` markers (in my opinion).

However, by sidestepping the problem, you continue building on non-robust legacy choice and will need to add increasingly complex regular expression(s) to hammer the non-ideal setup into a working solution. So, I agree with @Koooooo-7: Going the  `getAndRemoveConfig` hook, with a parser way, you get more flexible, more robust, and easier to maintain code.

jhildenbiddle commented:
@Jacco-V, @Koooooo-7 --

No arguments from me about the issues with building upon problematic legacy code. My comment was intended only to propose a way to make the stated drawback of this PR -- the inability to support  escaped characters in quoted-string config-values -- less of an issue so that we can delivery a solution sooner. If someone wants to take a crack at doing things the "better" way, I'm all for it.
